### PR TITLE
#16309 - Add api endpoints for Affiliation Invitation - request type

### PR DESCRIPTION
**Pull request by hfekete**
*Issue #:16309*
https://github.com/bcgov/entity/issues/16309

Adding API calls to support creation, and specific queries for Affiliation Invitations of type 'REQUEST'

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-auth license (Apache 2.0).

### CODE REVIEW

File: auth-api/src/auth_api/resources/affiliation_invitation.py [hunk 1]
```diff
@@ -163,3 +163,64 @@ def put(affiliation_invitation_id, affiliation_invitation_token):
         except BusinessException as exception:
             response, status = {'code': exception.code, 'message': exception.message}, exception.status_code
         return response, status
+
+
+@cors_preflight('GET,DELETE,OPTIONS')
+@API.route('/<string:affiliation_invitation_id>/authorize',
```
Comment (seeker25) [inline auth-api/src/auth_api/resources/affiliation_invitation.py:169]: Is there a reason why we're building so many routes for affiliation invitations? Can we slim these down and reuse what Ody already built?

File: auth-api/src/auth_api/resources/org.py [hunk 2]
```diff
@@ -322,6 +322,55 @@ def delete(org_id):
...
+    """Resource for managing affiliation invitations for an org."""
+
+    @staticmethod
+    def _verify_permissions(org_id):
+        if UserService.is_context_user_staff():
+            return
+
+        user = UserService.find_by_jwt_token()
+
+        if not user:
```
Comment (seeker25) [inline auth-api/src/auth_api/resources/org.py:337]: This shouldn't happen

File: auth-api/src/auth_api/resources/org.py [hunk 3]
```diff
@@ -322,6 +322,55 @@ def delete(org_id):
...
+
+        try:
+            cls._verify_permissions(org_id)
+
+            data = AffiliationInvitationService. \
+                get_all_invitations_with_details_related_to_org(org_id=org_id, statuses=filter_by_status)
+            data = [AffiliationInvitationService.as_dict_with_details
+                    (elem[0], elem[1], elem[2], elem[3]) for elem in data]
+
+            response, status = {'affiliationInvitations': data}, http_status.HTTP_200_OK
```
Comment (seeker25) [inline auth-api/src/auth_api/resources/org.py:366]: This looks similar to one of the routes Odysseus built

**Review by seeker25 — CHANGES_REQUESTED** [review]
Refer to the comments below

File: auth-api/src/auth_api/resources/entity.py [hunk 4]
```diff
@@ -214,3 +215,49 @@ def get(business_identifier):
...
+
+    @staticmethod
+    @_jwt.requires_auth
+    @cors.crossdomain(origin='*')
+    def get(business_identifier, org_id):
+        """Return list of Affiliation Invitations with details."""
+        status = request.args.get('status', None)
+        statuses = request.args.get('statuses', None)
+        affiliation_invitation_type = request.args.get('type', None)
+        filter_by_status = None
```
Comment (seeker25) [inline auth-api/src/auth_api/resources/entity.py:233]: I don't think these two are needed

File: auth-api/src/auth_api/models/affiliation_invitation.py [hunk 5]
```diff
@@ -17,18 +17,21 @@
...
+from auth_api.utils.enums import AffiliationInvitationType as AffiliationInvitationTypeEnum
 from .affiliation_invitation_type import AffiliationInvitationType
 
 from .base_model import BaseModel
 from .db import db
 from .invite_status import InvitationStatus
 
 
-class AffiliationInvitation(BaseModel):  # pylint: disable=too-many-instance-attributes
+class AffiliationInvitation(BaseModel):  # pylint: disable=too-many-instance-attributes,too-many-public-methods
```
[the remark this answers is not in the document]
Comment (hfekete) [reply auth-api/src/auth_api/models/affiliation_invitation.py]: Ok, did a bit of refactoring on the method, removed a lot of methods, and replaced them with single one.

File: auth-api/src/auth_api/resources/affiliation_invitation.py [hunk 6]
```diff
@@ -163,3 +163,64 @@ def put(affiliation_invitation_id, affiliation_invitation_token):
...
+    def _verify_permissions(user, affiliation_invitation_id):
+        if not user:
+            raise BusinessException(Error.NOT_AUTHORIZED_TO_PERFORM_THIS_ACTION, None)
+
+        affiliation_invitation = AffiliationInvitationService. \
+            find_affiliation_invitation_by_id(affiliation_invitation_id)
+        if not affiliation_invitation:
+            raise BusinessException(Error.DATA_NOT_FOUND, None)
+
+        to_org_id = affiliation_invitation.as_dict()['to_org']['id']
```
[the remark this answers is not in the document]
Comment (hfekete) [reply auth-api/src/auth_api/resources/affiliation_invitation.py:184]: lets leave it for now, I agree that dataclas would probably be better, but unfortunately currently its bit hard to replace everything with it.

File: auth-api/src/auth_api/resources/entity.py [hunk 7]
```diff
@@ -214,3 +215,49 @@ def get(business_identifier):
...
+                                                                          skip_auth=True)):
+                raise BusinessException(Error.DATA_NOT_FOUND, None)
+
+            data = AffiliationInvitationService. \
+                get_all_invitations_sent_to_org_for_entity(org_id=org_id,
+                                                           entity_id=business.identifier,
+                                                           status_filters=filter_by_status,
+                                                           types_filter=filter_by_types)
+
+            data = [AffiliationInvitationService.as_dict_with_details
```
[the remark this answers is not in the document]
Comment (hfekete) [reply auth-api/src/auth_api/resources/entity.py]: removed that code, I didnt like the custom stuff either.

File: auth-api/src/auth_api/services/affiliation_invitation.py [hunk 8]
```diff
@@ -406,3 +413,81 @@ def accept_affiliation_invitation(affiliation_invitation_id,
...
+    def as_dict_with_details(affiliation_invitation, from_org, to_org, entity):
+        """Return nested dict with details about affiliation invitation and sub entities."""
+        affiliation_invitation = AffiliationInvitation(affiliation_invitation).as_dict()
+        from_org = OrgService(from_org).as_dict()
+        to_org = OrgService(to_org).as_dict()
+        entity = EntityService(entity).as_dict()
+
+        msg = affiliation_invitation['additional_message'] if 'additional_message' in affiliation_invitation else ''
+
+        return {
```
[the remark this answers is not in the document]
Comment (hfekete) [reply auth-api/src/auth_api/services/affiliation_invitation.py]: I updated the code, but will stick with Marshmellow for now. I am open to change this stuff in the future (currently tight on time). I would not mind doing this in one of the next PRs

File: auth-api/src/auth_api/models/affiliation_invitation.py [hunk 9]
```diff
@@ -195,3 +204,52 @@ def expire_invitation(self):
         self.invitation_status = InvitationStatus.get_status_by_code(InvitationStatuses.EXPIRED.value)
         self.save()
         return self
+
+    def fail_invitation(self):
```
[the remark this answers is not in the document]
Comment (hfekete) [reply auth-api/src/auth_api/models/affiliation_invitation.py]: added `set_status` method

File: auth-api/src/auth_api/resources/affiliation_invitation.py [hunk 10]
```diff
@@ -163,3 +163,64 @@ def put(affiliation_invitation_id, affiliation_invitation_token):
...
+        if not UserService.is_user_member_of_org(user=user, org_id=to_org_id):
+            raise BusinessException(Error.NOT_AUTHORIZED_TO_PERFORM_THIS_ACTION, None)
+
+    @classmethod
+    @TRACER.trace()
+    @cors.crossdomain(origin='*')
+    @_jwt.requires_auth
+    def get(cls, affiliation_invitation_id: str):
```
[the remark this answers is not in the document]
Comment (hfekete) [reply auth-api/src/auth_api/resources/affiliation_invitation.py]: changed to PUT

File: auth-api/src/auth_api/resources/affiliation_invitation.py (hunk 1 above)
Comment (hfekete) [reply auth-api/src/auth_api/resources/affiliation_invitation.py:169]: I would prefer to have this as additional route as it allows me to add checks that might not be necessary for other routes. Like checking that person authorizing the invitation is actually part of the org.

File: auth-api/src/auth_api/resources/org.py (hunk 2 above)
Comment (hfekete) [reply auth-api/src/auth_api/resources/org.py]: Ok, removed that check

File: auth-api/src/auth_api/resources/org.py (hunk 3 above)
Comment (hfekete) [reply auth-api/src/auth_api/resources/org.py:361]: cleaned up code a little bit, though route might be similar I think they are doing different things in the service calls.

File: auth-api/src/auth_api/models/affiliation_invitation.py [hunk 11]
```diff
@@ -105,15 +103,58 @@ def create_from_dict(cls, invitation_info: dict, user_id, affiliation_id=None):
         affiliation_invitation.save()
         return affiliation_invitation
 
+    # pylint: disable=too-many-arguments
     @classmethod
-    def find_invitations_by_sender(cls, user_id):
-        """Find all affiliation invitations sent by the given user."""
-        return cls.query.filter_by(sender_id=user_id).all()
+    def filter_by(cls,
```
Comment (seeker25) [inline auth-api/src/auth_api/models/affiliation_invitation.py:108]: Use data class, this is way too many parameters.

File: auth-api/src/auth_api/resources/affiliation_invitation.py (hunk 1 above)
Comment (seeker25) [reply auth-api/src/auth_api/resources/affiliation_invitation.py:169]: and why couldn't those conditionally be built into the route that exists?

File: auth-api/src/auth_api/resources/org.py (hunk 3 above)
Comment (seeker25) [reply auth-api/src/auth_api/resources/org.py:361]: So why not change the route he built? so we don't have two different versions that manage the same entity?

File: auth-api/src/auth_api/models/affiliation_invitation.py [hunk 12]
```diff
@@ -89,22 +95,66 @@ def create_from_dict(cls, invitation_info: dict, user_id, affiliation_id=None):
...
+                  entity_id=None,
+                  affiliation_id=None,
+                  status_codes=None,
+                  invitation_types=None
+                  ) -> list:
+        """Filter list of Affiliation Invitations by provided filters. At least one filter needs to be provided."""
+        results = db.session.query(AffiliationInvitation)
+        filter_set = False
+
+        if from_org_id is not None:
```
Comment (seeker25) [inline auth-api/src/auth_api/models/affiliation_invitation.py:122]: `if from_org_id:`

File: auth-api/src/auth_api/schemas/affiliation_invitation.py [hunk 13]
```diff
@@ -29,8 +29,9 @@ class Meta(BaseSchema.Meta):  # pylint: disable=too-few-public-methods
         model = AffiliationInvitationModel
         fields = (
             'id', 'from_org', 'to_org', 'business_identifier', 'recipient_email', 'sent_date', 'expires_on',
-            'accepted_date', 'status', 'token', 'type', 'affiliation_id')
+            'accepted_date', 'status', 'token', 'type', 'affiliation_id', 'entity')
 
     from_org = fields.Nested('OrgSchema', only=('id', 'name', 'org_type'))
     to_org = fields.Nested('OrgSchema', only=('id', 'name', 'org_type'))
+    entity = fields.Nested('EntitySchema', only=('business_identifier', 'name'))
```
Comment (seeker25) [inline auth-api/src/auth_api/schemas/affiliation_invitation.py:36]: You can't rely on the name here - we're axing this data shortly.

File: auth-api/src/auth_api/resources/affiliation_invitation.py (hunk 1 above)
Comment (hfekete) [reply auth-api/src/auth_api/resources/affiliation_invitation.py:169]: I am reusing some (like creation of invitation), but I think that some actions are better pulled out in separate endpoints (like authorizing/not authorizing) to keep other endpoints more simple. I do try to keep a balance between increasing complexity on certain endpoints and adding new ones.

Is there a specific issue with having more endpoints?

File: auth-api/src/auth_api/schemas/affiliation_invitation.py (hunk 13 above)
Comment (hfekete) [reply auth-api/src/auth_api/schemas/affiliation_invitation.py:36]: Is it going away entirely from the table?
If so, how do I retrieve this data?

Comment (seeker25) [reply auth-api/src/auth_api/schemas/affiliation_invitation.py:36]:
```
business = AffiliationInvitation._get_business_details(business_identifier,
                                                                   RestService.get_service_account_token())                                                                   ```

Comment (seeker25) [reply auth-api/src/auth_api/schemas/affiliation_invitation.py:36]: If you're looking at querying multiple business_identifiers, look at `LEAR_AFFILIATION_DETAILS_URL` and trace it up

File: auth-api/src/auth_api/services/affiliation_invitation.py [hunk 14]
```diff
@@ -137,6 +151,7 @@ def create_affiliation_invitation(affiliation_invitation_info: Dict,
             # Fetch the up-to-date business details from legal API
             business = AffiliationInvitation._get_business_details(business_identifier,
                                                                    RestService.get_service_account_token())
+            # business = {'business': {'legalName': 'hello whatsup'}}
```
Comment (seeker25) [inline auth-api/src/auth_api/services/affiliation_invitation.py:154]: This should probably be removed

File: auth-api/src/auth_api/schemas/affiliation_invitation.py (hunk 13 above)
Comment (seeker25) [reply auth-api/src/auth_api/schemas/affiliation_invitation.py:36]: image

image

File: auth-api/src/auth_api/resources/affiliation_invitation.py [hunk 15]
```diff
@@ -163,3 +163,64 @@ def put(affiliation_invitation_id, affiliation_invitation_token):
...
+
+    @classmethod
+    @TRACER.trace()
+    @cors.crossdomain(origin='*')
+    @_jwt.requires_auth
+    def delete(cls, affiliation_invitation_id: str):
+        """Check if user is active part of the Org. Refuse invite authorization if he is."""
+        try:
+            user = UserService.find_by_jwt_token()
+            cls._verify_permissions(user=user, affiliation_invitation_id=affiliation_invitation_id)
```
[the remark this answers is not in the document]
Comment (seeker25) [reply auth-api/src/auth_api/resources/affiliation_invitation.py:219]: We already have a delete method 

image

Can we use patch instead?

image

File: auth-api/src/auth_api/resources/affiliation_invitation.py [hunk 16]
```diff
@@ -163,3 +163,66 @@ def put(affiliation_invitation_id, affiliation_invitation_token):
...
+
+        to_org_id = affiliation_invitation.as_dict()['to_org']['id']
+        if not UserService.is_user_member_of_org(user=user, org_id=to_org_id):
+            raise BusinessException(Error.NOT_AUTHORIZED_TO_PERFORM_THIS_ACTION, None)
+
+    @classmethod
+    @TRACER.trace()
+    @cors.crossdomain(origin='*')
+    @_jwt.requires_auth
+    def put(cls, affiliation_invitation_id: str):
```
Comment (seeker25) [inline auth-api/src/auth_api/resources/affiliation_invitation.py:192]: This is good to be on a route by itself

File: auth-api/migrations/versions/2023_08_01_ba5fafee24c1_.py [hunk 17]
```diff
@@ -0,0 +1,24 @@
+"""empty message
```
Comment (seeker25) [inline auth-api/migrations/versions/2023_08_01_ba5fafee24c1_.py:1]: Please reorder the migrations, so we don't need this. Ideally they'd be sequential as they're getting merged into main.

File: auth-api/src/auth_api/services/user.py [hunk 18]
```diff
@@ -639,3 +639,21 @@ def __remove_org_membership(org, user_id):
...
+    @user_context
+    def is_context_user_staff(**kwargs):
+        """Check if user in user context has is a staff."""
+        user_from_context: UserContext = kwargs['user_context']
+        return user_from_context.is_staff()
+
+    @staticmethod
+    def is_user_member_of_org(user, org_id: int) -> bool:
+        """Verify if user (UserService wrapper) is active member of the provided org."""
+        if not user:
```
[the remark this answers is not in the document]
Comment (seeker25) [reply auth-api/src/auth_api/services/user.py:653]: I think it's fine to leave this in, will make unit testing much easier

File: auth-api/src/auth_api/models/affiliation_invitation.py [hunk 19]
```diff
@@ -97,14 +105,53 @@ def create_from_dict(cls, invitation_info: dict, user_id, affiliation_id=None):
         return affiliation_invitation
 
     @classmethod
-    def find_invitations_by_sender(cls, user_id):
-        """Find all affiliation invitations sent by the given user."""
-        return cls.query.filter_by(sender_id=user_id).all()
+    def filter_by(cls, search_filter: AffiliationInvitationSearch, query=None) -> list:
```
Comment (seeker25) [inline auth-api/src/auth_api/models/affiliation_invitation.py:108]: Lookin good

File: auth-api/src/auth_api/services/affiliation_invitation.py [hunk 20]
```diff
@@ -406,3 +413,81 @@ def accept_affiliation_invitation(affiliation_invitation_id,
...
+            .find_all_sent_to_org_for_entity(to_org_id=org_id,
+                                             entity_id=entity_id,
+                                             status_filters=status_filters,
+                                             types_filter=types_filter)
+
+    @staticmethod
+    def refuse_affiliation_invitation(invitation_id):
+        """Set affiliation invitation to FAILED state (refusing authorization for affiliation)."""
+        invitation: AffiliationInvitationModel = AffiliationInvitationModel.find_invitation_by_id(invitation_id)
+        if not invitation:
```
[the remark this answers is not in the document]
Comment (seeker25) [reply auth-api/src/auth_api/services/affiliation_invitation.py:481]: `if not (invitation := AffiliationInvitationModel = AffiliationInvitationModel.find_invitation_by_id(invitation_id))`

File: auth-api/src/auth_api/services/affiliation_invitation.py [hunk 21]
```diff
@@ -159,6 +224,21 @@ def _get_business_details(business_identifier: str, token: str):
 
         return get_business_response.json()
 
+    @staticmethod
+    def _get_multiple_business_details(business_identifiers: List[str], token: str) -> List:
```
Comment (seeker25) [inline auth-api/src/auth_api/services/affiliation_invitation.py:228]: Excellent

File: auth-api/src/auth_api/resources/affiliation_invitation.py (hunk 1 above)
Comment (hfekete) [reply auth-api/src/auth_api/resources/affiliation_invitation.py:193]: Updated it. Now anything that does search can be done via GET on resource.

File: auth-api/src/auth_api/models/affiliation_invitation.py (hunk 11 above)
Comment (hfekete) [reply auth-api/src/auth_api/models/affiliation_invitation.py]: Added

File: auth-api/src/auth_api/schemas/affiliation_invitation.py (hunk 13 above)
Comment (hfekete) [reply auth-api/src/auth_api/schemas/affiliation_invitation.py]: Thanks for sharing this !

File: auth-api/migrations/versions/2023_08_01_ba5fafee24c1_.py (hunk 17 above)
Comment (hfekete) [reply auth-api/migrations/versions/2023_08_01_ba5fafee24c1_.py]: Done. 
What I did was:
rollback, manually edited migration files to point which one is before, did upgrade again. Had to clean up the data from the table though. Let me know if it looks good now.

Comment (seeker25) [reply auth-api/migrations/versions/2023_08_01_ba5fafee24c1_.py]: Looks great, that's what we usually do - pain in the neck but then the migrations are linear

**Review by seeker25 — APPROVED** [review]

**Pull request merged**